snorkel/slicing/utils.py

```diff
@@ -0,0 +1,202 @@
+from functools import partial
+from typing import List
+
+import numpy as np
+import torch
+from scipy.sparse import csr_matrix
+from torch import nn
+
+from snorkel.model.utils import convert_labels
+from snorkel.mtl.data import MultitaskDataLoader
+from snorkel.mtl.modules.utils import ce_loss, softmax
+from snorkel.mtl.scorer import Scorer
+from snorkel.mtl.task import Operation, Task
+
+from .modules.slice_combiner import SliceCombinerModule
+
+
+def add_slice_labels(
+    base_task: Task,
+    dataloader: MultitaskDataLoader,
+    slice_labels: csr_matrix,
+    slice_names: List[str],
+):
+    slice_labels = slice_labels.toarray()
+    slice_labels, slice_names = _add_base_slice(slice_labels, slice_names)
+
+    label_name = dataloader.task_to_label_dict[base_task.name]
+    labels = dataloader.dataset.Y_dict[label_name]
+    for i, slice_name in enumerate(slice_names):
+
+        # Convert labels
+        indicators = torch.LongTensor(slice_labels[:, i])
+        ind_labels = convert_labels(indicators, source="onezero", target="categorical")
+        pred_labels = indicators * labels
+
+        ind_task_name = f"{base_task.name}_slice:{slice_name}_ind"
+        pred_task_name = f"{base_task.name}_slice:{slice_name}_pred"
+
+        # Update dataloaders
+        dataloader.dataset.Y_dict[ind_task_name] = ind_labels
+        dataloader.dataset.Y_dict[pred_task_name] = pred_labels
+
+        dataloader.task_to_label_dict[ind_task_name] = ind_task_name
+        dataloader.task_to_label_dict[pred_task_name] = pred_task_name
+    return dataloader
+
+
+def convert_to_slice_tasks(base_task: Task, slice_names: List[str]):
+    """Adds slice labels to dataloader and creates new slice tasks (including base slice)
+
+    Each slice will get two slice-specific heads:
+    - an indicator head that learns to identify when DataPoints are in that slice
+    - a predictor head that is trained on only members of that slice
+
+    The base task's head is replaced by a master head that makes predictions based on
+    a combination of the predictor heads' predictions that are weighted by the
```

Comment:
nit: to be more precise, "prediction heads' predictions" >> "prediction heads' prediction confidences"